presto-main/src/main/java/com/facebook/presto/util/JsonUtil.java

```diff
@@ -1300,40 +1303,72 @@ public static void parseJsonToSingleRowBlock(
             for (int i = 0; i < fieldAppenders.length; i++) {
                 parser.nextToken();
-                fieldAppenders[i].append(parser, singleRowBlockWriter);
+                fieldAppenders[i].append(parser, singleRowBlockWriter, sqlFunctionProperties);
             }
             if (parser.nextToken() != JsonToken.END_ARRAY) {
                 throw new JsonCastException(format("Expected json array ending, but got %s", parser.getText()));
             }
         }
         else {
             verify(parser.getCurrentToken() == START_OBJECT);
-            if (!fieldNameToIndex.isPresent()) {
+            Optional<Map<Field, Integer>> fieldToIndex = getFieldToIndex(rowType.getFields());
+            if (!fieldToIndex.isPresent()) {
                 throw new JsonCastException("Cannot cast a JSON object to anonymous row type. Input must be a JSON array.");
             }
             boolean[] fieldWritten = new boolean[fieldAppenders.length];
             int numFieldsWritten = 0;
 
+            Map<String, Integer> caseSensitiveWhenMatching = new HashMap<>();
+            Map<String, Integer> caseInsensitiveWhenMatching = new HashMap<>();
+            if (sqlFunctionProperties.isLegacyJsonCast()) {
+                fieldToIndex.get().entrySet().stream()
+                        .forEach(entry -> caseInsensitiveWhenMatching.put(
+                                entry.getKey().getName().get().toLowerCase(ENGLISH),
+                                entry.getValue()));
+            }
+            else {
+                fieldToIndex.get().entrySet().stream()
+                        .forEach(entry -> {
+                            if (entry.getKey().isDelimited()) {
+                                caseSensitiveWhenMatching.put(
+                                        entry.getKey().getName().get(),
+                                        entry.getValue());
+                            }
+                            else {
+                                caseInsensitiveWhenMatching.put(
+                                        entry.getKey().getName().get().toLowerCase(ENGLISH),
+                                        entry.getValue());
+                            }
+                        });
+            }
             while (parser.nextToken() != JsonToken.END_OBJECT) {
                 if (parser.currentToken() != FIELD_NAME) {
                     throw new JsonCastException(format("Expected a json field name, but got %s", parser.getText()));
                 }
-                String fieldName = parser.getText().toLowerCase(Locale.ENGLISH);
-                Integer fieldIndex = fieldNameToIndex.get().get(fieldName);
+
+                String fieldName = parser.getText();
+                Integer fieldIndex = caseSensitiveWhenMatching.get(fieldName);
+                if (fieldIndex == null) {
+                    fieldIndex = caseInsensitiveWhenMatching.get(fieldName.toLowerCase(ENGLISH));
+                }
```

Comment:
for safety, throw an exception if it's not in either.

Reply:
Sure!

Reply:
After re-check the code, I think maybe we shouldn't throw an exception when it's not in either, so that it would fall into the `parser.skipChildren()` below. The result is, when we have a field in json that is not defined and matched in the row type to cast, we would ignore the field rather than throw an exception. For example, executing the following statement:

```
SELECT CAST(JSON '{"v1":123,"v2":"abc","v3":true, "v4":"1234"}' AS ROW(v1 BIGINT, v2 VARCHAR, v3 BOOLEAN, v5 VARCHAR));
```

we would get the following result, rather than get an exception:

```
               _col0                
------------------------------------
 {v1=123, v2=abc, v3=true, V5=null} 
```
What do you think, is this behavior reasonable?